SERVER-24 Apple Music API 관련 엔티티 작성

jher235 opened the pull request:
## 🔊 Summaries
<!-- 본인이 한 작업을 요약해주세요. -->
- Apple Music API 로부터 의존성을 분리하기 위해 Confeti 에서 데이터를 이중으로 관리하고자 하였고 이를 위한 엔티티를 작성햇습니다.

## ✨ Notification 
<!-- 참고할 사항을 작성해주세요. -->
- 원래는 기존 코드를 유지하며 연관관계 매핑도 진행하려했는데, 칼럼명에 중복이 생기는 부분이 존재하여 매핑 작업은 엔티티 확인 후 진행하려고 합니다.
- Song과 Music 중 네이밍을 Song로 둔 이유는 Apple api에서는 Music을 좀 더 포괄적인 개념으로 취급하고 있고 그 하위에 노래 (== Song)으로 관리하고 있는 듯 하여 네이밍을 Song으로 두었습니다.
- String 타입으로 관리하는 artist_id, song_id 값은 따로 길이 제한을 두진 않았습니다. MySQL 기준 기본값으로 varchar 255 이고, 255 이하는 길이를 저장하는 공간의 크기가 모두 동일하기 때문에 255 보다 짧게 두는 것에 대한 이점이 없다고 생각했습니다. 그리고 기존에 관리 중인 artist_id, song_id 등의 데이터를 확인해보니 해당 부분들이 현재는 숫자 데이터만 존재하는 듯하여 Long으로 변경하는 것이 좋을지, 혹은 문자열이 필요한 값들도 존재하는지 궁금합니다. 
- url의 경우 길이 제한에 걸리지 않도록 값을 좀 널널하게 두기 위해 length = 3000 으로 두었습니다.
- TopArtist 의 경우 PK를 rank로 두었습니다

### 연관관계 매핑 시 수정해야 하는 기존 엔티티 목록은 다음과 같습니다

- Artist를 추가해야하는 엔티티
1. `ArtistFavorite`
2. `ConcertArtist`
3. `FestivalArtist`
4. `PerformanceArtist`

- Song을 추가해야하는 엔티티
1. `SetlistMusic`

jher235 committed:
SERVER-24 feat: Artist 엔티티 및 JpaRepository 생성

jher235 committed:
SERVER-24 feat: Song 엔티티 및 JpaRepository 생성

jher235 committed:
SERVER-24 feat: ArtistSong 엔티티 및 JpaRepository 생성

jher235 committed:
SERVER-24 feat: TopArtist 엔티티 및 JpaRepository 생성

jher235 committed:
SERVER-24 fix: 잘못된 제네릭 타입 수정

jher235 committed:
SERVER-24 chore: 패키지 구조를 infra 내부에 repository로 수정

jher235 committed:
SERVER-24 feat: 아이디를 인조 식별자로 변경
- rank를 id로 둬도 큰 문제는 없겠지만, 프로덕트의 다른 테이블들과 통일해서 사용할 수 있도록 인조식별자를 pk로 사용하도록 변경함

jher235 committed:
SERVER-24 chore: 테이블명 복수형으로 통일

noeyag commented:
LGTM 입니다 🚀🚀🚀🚀🚀

ch1hyun commented on src/main/java/org/sopt/confeti/domain/artist/Artist.java:
```diff
@@ -0,0 +1,53 @@
+    private Integer artworkWidth;
+
+    private Integer artworkHeight;
+
```
created_at, updated_at 컬럼은 다른 엔티티 참고해서 필수로 만들어주세요

ch1hyun commented on src/main/java/org/sopt/confeti/domain/artist/Artist.java:
```diff
@@ -0,0 +1,53 @@
+    @Column(nullable = false)
+    private String name;
+
+    @Column(length = 3000)
```
길이를 3000자로 설정하신 이유가 있으신가요?

jher235 replied:
예전에 Internet Explorer 에서 최대 길이가 2083 자였는데 현재에 와서는 최대 길이 제한이 3만자 이상으로 늘어났지만 해당 경우는 거의 없기 때문에 2083자를 넘는 3000자 정도로 세팅해두었습니다.

ch1hyun commented on src/main/java/org/sopt/confeti/domain/artist/Artist.java:
```diff
@@ -0,0 +1,53 @@
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+public class Artist {
+
+    @Id
+    @Column(name = "artist_id")
+    private String id;
```
long 타입의 pk를 두는 것이 인덱스 관점에서 성능이 더 좋을 것 같습니다. 재헌님 의견은 어떠신가요?

jher235 replied:
저도 동일하게 생각하는데, 지난번에 치현님 말씀듣고 애플 문서를 확인해보니까 String으로 제시하고 있더라구요... 추후 문자열이 섞인 값이 오면 수정이 좀 번거롭진 않을까요??

ch1hyun replied:
엇 제 의견은 `artist_id` 는 `String` 타입으로 따로 두고, PK를 `Long` 타입의 `id`로 추가하자는 거였습니다!

jher235 replied:
아하 좋은 방법인 것 같습니다. 그런데 혹시 이렇게 lond type pk를 둔 경우 apple music 의 artist id로 조회해야만 하는 경우와 저희 서비스에서 자체적으로 저장해둔 artist의 long 타입 pk로 조회해야 하는 경우의 비율이 어느정도일까요? 
만약 대부분의 경우 apple music id 를 사용해야 한다면 불필요한 칼럼만 증가할까봐 걱정이 됩니다

jher235 replied:
제가 좀 더 고민해봤는데, 저희가 페스티벌이나 콘서트 등의 정보를 넣을 때 아티스트를 따로 지정해주는 부분들이 많아서 Long 타입으로 id를 넣어주는게 더 좋을 것 같다는 생각이 들었어요.
String도 사실 그렇게 길지 않아서 보통 11바이트 내로 처리될 것 같고, 쓰기 작업이 별로 없을 것 같다는 생각에 괜찮겠다는 생각도 존재하지만 역시 꼭 apple music api의 id값으로만 탐색할 필요가 없다면 이렇게 추가하는게 좋을 것 같습니다. 
Long 타입 ID를 추가하도록 작업할까요?

ch1hyun replied:
좋습니다! 그러면 `artist_id`를 `Long` 타입으로 저장하고, PK로 지정해주세요

ch1hyun commented on src/main/java/org/sopt/confeti/domain/artist_song/ArtistSong.java:
```diff
@@ -0,0 +1,52 @@
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(name = "artist_song_id")
```
각 테이블의 아이디 컬럼은 `id` 로 이름이 통일되어있어요! 이건 컨벤션이라 수정부탁드립니다.

ch1hyun commented on src/main/java/org/sopt/confeti/domain/artist_song/ArtistSong.java:
```diff
@@ -0,0 +1,52 @@
+    @ManyToOne(fetch = FetchType.LAZY, optional = false)
+    @JoinColumn(name = "song_id")
+    private Song song;
+
```
created_at, updated_at 추가해주세요

ch1hyun commented on src/main/java/org/sopt/confeti/domain/song/Song.java:
```diff
@@ -0,0 +1,52 @@
+
+    @Id
+    @Column(name = "song_id")
+    private String id;
```
아티스트와 동일합니다!

ch1hyun commented on src/main/java/org/sopt/confeti/domain/top_artist/TopArtist.java:
```diff
@@ -0,0 +1,35 @@
+    @JoinColumn(name = "artist_id")
+    private Artist artist;
+
+    @Column(nullable = false)
```
유니크 옵션도 추가하면 좋을 것 같아요

ch1hyun reviewed (changes requested):
고생하셨습니다!
현재 아티스트 엔티티 파일이 domain 패키지 하위에 있는 것 같습니다.
저희 비지니스 도메인과 분리하기 위해 `domain/applemusic` 패지키를 생성하고 하위로 옮겨주세요.
다른 패키지의 경우 띄어쓰기가 있을 때 `_`로 표시되어있는데, `_`를 없애기 위해 리펙토링할 예정입니다.

ch1hyun commented on src/main/java/org/sopt/confeti/domain/top_artist/TopArtist.java:
```diff
@@ -0,0 +1,35 @@
+
+    @Column(nullable = false)
+    private int rank;
+
```
아 여기 빌더, static 생성자가 없네요

jher235 committed:
SERVER-24 fix: id 칼럼명 수정

jher235 committed:
SERVER-24 fix: createdAt, updatedAt 공통 필드 추가

jher235 committed:
SERVER-24 feat: 생성자 추가 및 unique 제약 조건 추가

jher235 committed:
SERVER-24 chore: 패키지 구조를 domain 하위의 applemusic으로 수정

ch1hyun commented on src/main/java/org/sopt/confeti/domain/applemusic/artist/Artist.java:
```diff
@@ -0,0 +1,60 @@
+import org.springframework.data.annotation.CreatedDate;
+import org.springframework.data.annotation.LastModifiedDate;
+
+@Entity
```
CreatedDate, LastModified 어노테이션이 동작하기 위해서는 이벤트 리스너를 등록해줘야해요
`@EntityListeners(AuditingEntityListener.class)`

jher235 replied:
헉 실수가 너무 잦네요... 죄송합니당 바로 반영하겠습니다~!

jher235 committed:
SERVER-24 fix: 누락된 이벤트 리스너 등록

ch1hyun commented on src/main/java/org/sopt/confeti/domain/applemusic/artist_song/ArtistSong.java:
```diff
@@ -0,0 +1,64 @@
+package org.sopt.confeti.domain.applemusic.artist_song;
```
`_` 제거 부탁드릴게요!

ch1hyun commented on src/main/java/org/sopt/confeti/domain/applemusic/top_artist/TopArtist.java:
```diff
@@ -0,0 +1,60 @@
+package org.sopt.confeti.domain.applemusic.top_artist;
```
여기도 부탁드립니다!

jher235 committed:
SERVER-24 chore: 패키지명 자바 컨벤션에 통일

jher235 committed:
SERVER-24 feat: 별도의 BIGINT 타입 인조식별자 추가
- String을 PK로 사용하는 경우 인덱스 크기가 조금 더 커질 수 있고 인덱스를 통한 값 비교에서 숫자형 데이터 타입이 우월하므로 현재 서비스에서 사용할 별도의 인조식별자를 클러스터링 키로 설정함.

jher235 committed:
SERVER-24 feat: ranking 필드 네이밍 변경
- SQL 예약어인 rank와 충돌하므로 변경

ch1hyun commented on src/main/java/org/sopt/confeti/domain/applemusic/song/Song.java:
```diff
@@ -0,0 +1,68 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    private String songId;
+
+    @Column(nullable = false)
+    private String name;
```
길이 지정 안하신 이유가 있으실까요!?

jher235 replied:
PR 에도 작성해두긴 했는데 varchar를 사용하면서 길이가 255를 넘어가지 않을 것 같고, 딱히 길이에 대한 도메인 룰이 존재하지 않은 경우에 대해선 길이 지정이 불필요하다고 생각했습니다. 
혹시 길이 지정을 하는게 좋다고 생각하세요? 있다면 하는 것에 대한 이점이 있나요??

ch1hyun replied:
메모리 절약인데, 데이터 양이 많지 않기도 해서 길이 지정 안해도 괜찮을 것 같네요~ 애플 문서에 명시된 것도 없으니까요!

ch1hyun commented on src/main/java/org/sopt/confeti/domain/applemusic/song/Song.java:
```diff
@@ -0,0 +1,68 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    private String songId;
```
notnull 추가해서 명시해줘도 좋을 것 같아요

jher235 replied:
유니크 제약조건도 걸어둘까요? 혹은 따로 DB단에서 따로 인덱스를 거는 방법을 선호하세요?

ch1hyun replied:
넵 유니크 걸어주세요! 인덱스는 DB 단에서 따로 거는걸 선호합니다
JPA에서도 인덱스를 생성할 수 있나요?

jher235 replied:
넹 @Table 어노테이션으로 걸어줄 수 있어요. 그런데 보통 운영환경에서는 ddl-auto를 꺼두기 때문에 DB에 직접 거는게 많은 것 같아요. 제 기억으로 `@Column(unique=true)` 와 같이 둘 경우 유니크 인덱스가 자동으로 생성되어서 이렇게 둘 바엔 차라리 유니크 인덱스를 직접 컨트롤하는게 나을지 여쭤본거였어요. 
유니크 인덱스를 @Table 로 명시할까요? 아니면 그냥 `@Column(unique=true)` 만 둘까요?

ch1hyun replied:
흠 DB에 직접 거는게 나을 것 같습니다! 아니면... `@table` 방식이 좋을 것 같은데 어떻게 생각하시나요?
찾아보니까 Column에 명시하는건 이름이 예측 불가라 별로더라고요

jher235 replied:
맞아요.. 그럼 우선 `@Table` 에서 작성하는 방식으로 작업해두겠습니다~

jher235 committed:
SERVER-24 feat: 비즈니스 키를 유니크 인덱스로 지정
